[ian.lock 양정일] 지하철 서비스 1~3단계 PR

Jjungs7 opened the pull request:
안녕하세요 이안입니다! 잘 부탁드립니다 🙏
이전에 진행했던 미션들 덕분에 구조는 어느정도 잡히는 것 같지만 아직 애매한 부분이 조금 남아있는 것 같아요
궁금한 사항들은 코드 중간에 코멘트로 남기겠습니다!

---
## 기능 목록
### 인증(auth)
- AuthenticationPrincipal 로 로그인 유저의 정보를 받을 수 있음. isThrow 인자로 유효하지 않을 회원일 경우 바로 UNAUTHORIZED를 반환할 수 있음
### 즐겨찾기(favorite)
- 회원만 즐겨찾기 기능을 사용할 수 있음
- CRD 작업 지원
### 요금 조회(path)
- 경로, 요금을 계산해주는 Path 도메인
- 사람을 나타내는 Person 도메인(로그인 정보를 넘겨주면 연령대를 받을 수 있음)
- 요금 계산을 위해 extraFare정보를 담은 SectionEdge
- 최단경로를 찾아줄 SubwayGraph wrapper class
- 비회원도 요금을 조회할 수 있음

Jjungs7 committed:
feat(subwaygraph): 노선도 최단경로 조회 추가

Jjungs7 committed:
feat(path): 최단경로 조회 웹 컨트롤러

- DB에서 Section에서 전체 노선 불러오는 기능 추가
- Path 서비스 및 컨트롤러 추가

Jjungs7 committed:
feat(auth): 사용자 인증 및 토큰 발급

- 서비스에 등록된 회원이 아닐 경우 UNAUTHORIZED를 반환

Jjungs7 committed:
feat(favorite): 즐겨찾기 기능 추가

- 인증된 회원만 즐겨찾기 기능을 사용할 수 있음
- 단위 테스트 추가

Jjungs7 committed:
feat(요금): 요금 계산 기능 추가

- 사람을 나타내는 Person 클래스 추가
- 경로에 따른 추가 요금 기능 구현

Jjungs7 committed:
refactor(auth): Principal의 예외 처리 방법 변경

- null체크 하는 부분을 Principal에 위임

Jjungs7 committed:
test(path): 도메인에 대한 테스트 추가

Jjungs7 committed:
refactor(요금): 너무 많은 일을 담당하던 줄을 여러 줄로 나눔

Jjungs7 commented on src/main/java/subway/line/application/LineService.java:
```diff
@@ -76,17 +77,43 @@ public void addLineStation(Long lineId, SectionRequest request) {
+        insertSectionsAndSet(line);
+    }
+
+    private void deleteLineAndSet(Long id) {
+        lineDao.deleteById(id);
+        PathService.newlyUpdated();
+    }
```
저희가 페어프로그래밍하면서 의견이 제일 많이 부딪힌 곳이에요. 먼저 배경을 말씀드리자면 최단경로 조회 요청이 들어올때마다 매번 graph를 만드는 것은 비효율적이니까 그래프를 캐싱하기로 하였습니다 (참고코드: [PathService.isUpdated](https://github.com/next-step/spring-subway-service-kakao/pull/25/files#diff-d8d15695c69d475e8ef30eee394b5d8a12ca8e125fe2af848235ee214047fa7dR31)). isUpdated 변수는 Section, Line에 insert, delete 작업이 일어날때마다 새로운 그래프를 생성해야한다는것을 알려주는 flag인데요, 이걸 어디서 세팅해주는지에 대한 의견이 두 개로 나뉘었습니다.
- 객체지향적으로 봤을때 Dao는 그래프가 바뀌었다는 것을 알리는 것은 이상하고 위 코드처럼 Service에서 delete, insert한 후에 flag를 세팅해야한다
  - _의견_: flag 설정은 Service가 담당해야 할 일이다
  - 단점: Section, Line을 건드리는 Service가 많아질 수록 일일이 개발자가 처리해줘야한다
- 개발자가 직접 작성해야하는 코드는 error prone하기 때문에 애초에 Dao 에서 flag를 설정해줘야한다.
  - _의견_: Dao에서 하는 것이 맞는지 의문이 든다
  - 장점: 새로운 Service가 생겨도 flag를 신경 쓸 필요가 없다
현재는 객체지향을 배우고 있기 때문에 첫 번째 의견으로 작성한 상태인데요, 어떻게 하는 것이 더 맞았을지 궁금합니다!

htrucci replied:
구현하신게 어떻게보면 Cache서비스 같은 부분이군요?! 🤔 
저도 고민이 많이 되는 부분인데요.. 정답은 없을 것 같은데, 어려운 주제군요 ^^;;;
제가 구현했다면 전자를 선택했을 것 같습니다. 비즈니스 로직에서 구현하는 것이 더 많은 케이스 관리를 가능하게 할 것 같아서요. 그냥 캐시서비스라고 생각한다면 기능에 따라 캐시를 날려야할수도 그냥 둘수도 있기떄문에,, 
states를 계속 분기로 가지고 다녀야하니, 본 케이스에 도움이 될지 [스테이트패턴](https://rongscodinghistory.tistory.com/139) 내용을 한번 살펴봐주세요 😄 
후자의 경우 flag를 비즈니스로직에서 관리해주지 않아도 되기때문에 확실히 어플리케이션이 단순해지는 장점이 있어, 비즈니스 레이어가 복잡해질 가능성이 낮다면 후자를 선택했을 것 같습니다.

Jjungs7 commented on src/main/java/subway/path/domain/SectionEdge.java:
```diff
@@ -0,0 +1,39 @@
+
+    @Override
+    protected double getWeight() {
+        return distance + fare / 10000.0;
```
저희가 사용하려는 다익스트라 라이브러리가 가중치를 double형 밖에 받지 않아서 약간의 꼼수를 사용했습니다. 저희의 의도는 먼저 최단거리에 해당하는 구간을 구하고, 만약 최단거리가 같은 구간이 여러 개라면 fare를 기준으로 반환하게끔 구현하였습니다!

htrucci commented on src/main/java/subway/AuthenticationPrincipalConfig.java:
```diff
@@ -1,28 +1,32 @@
-        return new AuthenticationPrincipalArgumentResolver(authService);
+        return new AuthenticationPrincipalArgumentResolver(memberService, jwtTokenProvider);
     }
 }
```
디폴트 코드였긴하지만, Config, PageController 등도 새로운 패키지 안으로 넣고
java.subway안에는 SubwayApplication만 남겼으면 좋겠습니다~! 😄

htrucci commented on src/main/java/subway/AuthenticationPrincipalConfig.java:
```diff
@@ -1,28 +1,32 @@
-    public void addArgumentResolvers(List argumentResolvers) {
+    public void addArgumentResolvers(List<HandlerMethodArgumentResolver> argumentResolvers) {
         argumentResolvers.add(createAuthenticationPrincipalArgumentResolver());
     }
```
일부러 제너릭 명시해준 부분인 것 같네요 👍

htrucci commented on src/main/java/subway/auth/application/AuthService.java:
```diff
@@ -1,4 +1,42 @@
+    public AuthService(JwtTokenProvider jwtTokenProvider, MemberDao memberDao) {
+        this.jwtTokenProvider = jwtTokenProvider;
+        this.memberDao = memberDao;
+    }
```
생성자를 통한 주입시 @Autowired를 명시하도록 스프링에서 가이드 하고 있습니다.
단일 생성자일때는 안해도 된다는데, 전 습관적으로 하고 있습니다 😄

htrucci commented on src/main/java/subway/exceptions/AuthorizationException.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.UNAUTHORIZED)
+public class AuthorizationException extends RuntimeException {
+    public AuthorizationException() {
+    }
+
+    public AuthorizationException(String message) {
+        super(message);
+    }
+}
```
맞게 잘 하셨지만 굳이 하나를 말씀드린다면,,
인증(Authentication) & 인가(Authorization) 의 차이를 명확히 인지하고 있는 것이 필요합니다.
실무에서 위 개념을 명확하게 구분짓지 않아 혼동을 겪었던 경험이 있었습니다 😄
인증은 로그인이 되었는가?, 인가는 권한이 있는가? 의 차이가 있기 때문에 로그인이 실패된 경우
UnauthenticatedException 으로 명명해주는것이 좀 더 정확할 것 같습니다.
토큰이 유효하지 않은 경우도 인가실패 보다는 인증실패에 가깝습니다. 🤔
참고 부탁드립니다 😄

Jjungs7 replied:
보안과 관련된 인플도교 영상을 봤을때 인증, 인가 이야기를 했던 것 같네요... 그 사이에 까먹었나봐요 😂 매번 봐도 헷갈리는 것 같아요
http code는 unauthorized인데 exception은 Unauthenticated인 것도 한 몫 하는 것 같아요. 그래도 이번 기회에 확실히 외울 수 있을 것 같아요!

htrucci replied:
httpstatus의 경우 401 Unauthorized로 통일되어 사용하긴 하지만,,
두 개념이 다르다는 부분만 잘 인지하고 계시면 될 것 같네요 😄

htrucci commented on src/main/java/subway/auth/application/AuthService.java:
```diff
@@ -1,4 +1,42 @@
+        }
+
+        String token = jwtTokenProvider.createToken(request.getEmail());
+        return new TokenResponse(token);
```
크게 중요하지는 않지만,,
요즘은 계정통합, 소셜계정, 멀티계정 등으로 이메일이 중복되는 경우도 있어서 (물론 지하철 서비스는 유일하지만 ^^;;)
email말고 PK인 member.id 를 이용하여 토큰을 생성하는건 어떨까요?

htrucci replied:
토큰을 조작할일이 없다면 바로 리턴해도 되겠어요 😄 
return TokenResponse.of(jwtTokenProvider.createToken(request.getEmail());

htrucci commented on src/main/java/subway/auth/application/AuthService.java:
```diff
@@ -1,4 +1,42 @@
+        }
+
+        return member.getEmail().equals(email) && member.getPassword().equals(password);
+    }
```
email만으로 db에서 password값을 가져온 후 비교하게 되면 뭐,, 가져온 password hash가 로그에 찍힐수도있고,,
무심코 보안이슈가 생길 수 있으니, 쿼리에서 아예 패스워드까지 비교하는 방법은 어떨까요?

htrucci commented on src/main/java/subway/auth/domain/AuthenticationPrincipal.java:
```diff
@@ -9,4 +9,6 @@
 public @interface AuthenticationPrincipal {
     boolean required() default true;
+
+    boolean isThrow() default true;
```
required를 썼어도 될 것 같은데, isThrow를 생성하여 사용한 이유가 궁금합니다 🤔

Jjungs7 replied:
required가 내부적으로 필요한 메소드같아보여서 새로 만들었어요. 근데 다시보니까 intelliJ 가 회색으로 안 쓰여지고 있다고 힌트를 주고 있었네요. 다음부턴 무엇을 의미하는지 좀 더 찾아보고 써보겠습니다!

htrucci commented on src/main/java/subway/auth/ui/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -5,25 +5,39 @@
+            return null;
+        }
+
+        return memberService.findMemberByEmail(jwtTokenProvider.getPayload(token));
```
indent를 1단계로 줄여볼까요? 🤔 
또한 null을 반환하지 않는 것이 좋아보입니다.

htrucci replied:
member.id로 회원정보를 가져오는것이 더 정확하겠죠? 🤔

htrucci commented on src/main/java/subway/auth/ui/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -5,25 +5,39 @@
+        if (!jwtTokenProvider.validateToken(token)) {
+            AuthenticationPrincipal auth = parameter.getParameterAnnotation(AuthenticationPrincipal.class);
+            if (auth.isThrow()) {
+                throw new AuthorizationException("");
```
Exception발생시 메시지를 비워두지않고 넣어주는게 좋지 않을까 생각이 됩니다. 😄

htrucci commented on src/main/java/subway/exceptions/ExceptionHandlers.java:
```diff
@@ -0,0 +1,15 @@
+    @ExceptionHandler(NoSuchElementException.class)
+    public ResponseEntity<Void> noElement() {
+        return ResponseEntity.badRequest().build();
+    }
```
ResponseEntity<Void> 를 반환하지 않고 @ResponseStatus를 적용해도 좋을 것 같네요!

htrucci commented on src/main/java/subway/exceptions/FavoriteDuplicateException.java:
```diff
@@ -0,0 +1,15 @@
+    public FavoriteDuplicateException(String message) {
+        super(message);
+    }
+}
```
👍

htrucci commented on src/main/java/subway/exceptions/FavoriteSameArgumentException.java:
```diff
@@ -0,0 +1,15 @@
+    public FavoriteSameArgumentException(String message) {
+        super(message);
+    }
+}
```
👍

htrucci commented on src/main/java/subway/favorite/application/FavoriteService.java:
```diff
@@ -0,0 +1,42 @@
+
+        if (favoriteDao.favoriteExists(favorite)) {
+            throw new FavoriteDuplicateException();
+        }
```
Exception메시지 넘겨준다면 디버깅에 더 용이하지 않을까 싶어요~!

htrucci commented on src/main/java/subway/favorite/dao/FavoriteDao.java:
```diff
@@ -0,0 +1,80 @@
+                "WHERE F.member_id = ?";
+
+        return jdbcTemplate.query(sql, favoriteMapper, memberId);
+    }
```
Dao에서는 Dto반환이 아닌 Favorite 원형으로 반환하고 서비스 로직에서 필요에 따라 Dto변환이 일어나도록 해주세요~! 😄

htrucci commented on src/main/java/subway/favorite/dao/FavoriteDao.java:
```diff
@@ -0,0 +1,80 @@
+                    new StationResponse(
+                            rs.getLong("target_station_id"),
+                            rs.getString("target_station_name"))
+            );
```
👍

htrucci commented on src/main/java/subway/favorite/dao/FavoriteDao.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    public long insert(Favorite favorite) {
+        String sql = "insert into favorite(member_id, source_station_id, target_station_id) values(?, ?, ?)";
```
테이블이 대문자로 생성되었다면 쿼리에서도 case맞추어 대문자로 표현해주는 것이 좋습니다! 😄

htrucci commented on src/main/java/subway/favorite/dao/FavoriteDao.java:
```diff
@@ -0,0 +1,80 @@
+    public boolean favoriteExists(Favorite favorite) {
+        String sql = "select EXISTS" +
+                " (select * from favorite where member_id = ? and source_station_id = ? and target_station_id = ?)" +
+                " as success";
```
EXISTS만 대문자로 작성이 되었네요 😄

htrucci replied:
EXISTS를 사용하기 위해 메인쿼리를 서브쿼리 구조로 넣는것이 필요할까? 🤔  생각이 듭니다.
카운트를 사용하여 판단하면 쿼리를 더 단순하게 만들 수 있지 않을까요?
select count(*) from favorite where member_id = ? and source_station_id = ? and target_station_id = ?

htrucci commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,51 @@
+        synchronized (PathService.class) {
+            isUpdated = false;
+        }
+    }
```
별도의 States 객체로 만들어 관리하면 어떨까요?

htrucci commented on src/main/java/subway/path/domain/Person.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.stream.Stream;
+
+public enum Person {
+    ADULT(0, 0, 19, Person.NO_LIMIT),
+    TEEN(20, 350, 13, 19),
+    CHILD(50, 350, 6, 13),
+    NEWBORN(100, 0, 0, 6);
+
+    private static final int NO_LIMIT = 100000;
+    private final int discountPercentage;
+    private final int deduction;
+    private final int minAge;
+    private final int maxAge;
```
💯

htrucci commented on src/main/java/subway/path/ui/PathController.java:
```diff
@@ -1,5 +1,30 @@
+                                                         @RequestParam("target") long targetId) {
+        Path path = pathService.find(sourceId, targetId);
+        int fare = pathService.calculateFare(path, Person.of(loginMember));
+        return ResponseEntity.ok(PathResponse.of(path, fare));
```
구간요금과 연령요금 체계를 잘 분리해서 구현해주셨네요 👍

htrucci commented on src/main/java/subway/path/domain/Person.java:
```diff
@@ -0,0 +1,48 @@
+
+    private boolean isInAge(LoginMember loginMember) {
+        return minAge <= loginMember.getAge() && loginMember.getAge() < maxAge;
+    }
```
비로그인 회원도 연령을 입력해서 요금 계산이 가능해야한다면? 🤔 
연령별 요금 계산에 있어 age만 필요하기때문에 LoginMember를 객체로 받지 않아도 되지 않을까요?

htrucci commented on src/main/java/subway/member/ui/MemberController.java:
```diff
@@ -42,19 +43,19 @@ public ResponseEntity createMember(@RequestBody MemberRequest request) {
     }
```
비로그인 상태에서도 타인의 계정을 지우거나 계정정보를 볼 수 있습니다.
본인 요청만 처리하도록 보안처리가 필요하지 않을까요?

Jjungs7 replied:
이 기능이 왜 있었는지 생각해보았는데 관리자 기능인 것 같아요. 현재는 레포지토리 의도에 맞게 인증된 회원만 기능을 사용할 수 있도록 변경하였습니다!

htrucci reviewed (changes requested):
안녕하세요, 지하철 서비스 리뷰를 하게 된 데빈입니다.
모든 기능이 정상적으로 동작하고 아주 잘 구현해주셨습니다 👍 👍 
몇가지 개선하면 좋을 사항에 대한 코멘트 남겼습니다. 
리뷰 남겨드린 내용 한번 고민해보시고 다시 PR 부탁드립니다~! 😄

Jjungs7 committed:
refactor(configs): 패키지 분리

- 각자 역할에 맞는 패키지로 이동

Related:
- https://github.com/next-step/spring-subway-service-kakao/pull/25#discussion_r564982511

Jjungs7 committed:
refactor: autowired, jsoncreator, final 및 람다식 형식 통일

- @Autowired를 명시하여 혼돈을 줄이기
- final로 선언해도 되는 필드는 final로 선언
- 빈 생성자를 최대한 방지하기 위하여 JsonCreator 사용
  - DTO 외에 다른 곳에서 사용된 JsonCreator는 삭제
- 불필요하게 선언된 isThrow() 메소드 삭제

**주의**: 테스트를 intelliJ를 사용할 경우 Java compiler parameter에 -parameters 옵션 추가해야함
Related:
- https://github.com/next-step/spring-subway-service-kakao/pull/25#discussion_r564983383
- https://github.com/next-step/spring-subway-service-kakao/pull/25#discussion_r564992519
- https://github.com/next-step/spring-subway-service-kakao/pull/25#discussion_r564993907

Jjungs7 committed:
refactor(exception): null 리턴 안하기, exception에 대한 정보 추가

- 용이한 추적을 위해 exception이 났을 때를 대비한 message 추가

Jjungs7 committed:
refactor: 불필요하게 많은 Line 생성자 개수 축소

- Line의 VO 역할을 LineAttributes에게 부여

Jjungs7 committed:
refactor(auth): LoginMember은 null일 수 없음, 유저 쿼리 방식 변경

- 유저를 찾을때 member의 비밀번호는 가져오지 않게 수정
- token은 id로 만들게 변경
- 회원가입하지 않은 유저지만 게스트 상태로 사용할 수 있는 기능에 대해서는 guestMember 권한 적용

Jjungs7 committed:
refactor(favorite): Dao에서 response가 아닌 즐겨찾기를 반환하도록 수정

- Favorite 에서 DTO 변환은 Service에서 이루어짐

Related:
- https://github.com/next-step/spring-subway-service-kakao/pull/25#discussion_r564997882

Jjungs7 committed:
refactor(person): Person 클래스가 나이만 받도록 수정

- age 정보만 필요한데 로그인정보 전체를 넘길 필요가 없음

Related:
- https://github.com/next-step/spring-subway-service-kakao/pull/25#discussion_r565024187

Jjungs7 committed:
refactor(dao): sql문에서 테이블명 대문자로 변경을

- 스키마와 일관성을 유지

Related:
- https://github.com/next-step/spring-subway-service-kakao/pull/25#discussion_r564998219

Jjungs7 committed:
fix(exception): 처리 안 되던 익셉션 추가

Jjungs7 committed:
refactor(member): 로그인된 회원만 기능을 사용할 수 있게 변경

Jjungs7 committed:
refactor(graph): 지하철 그래프에 스테이트패턴 적용

- 지하철 그래프는 최신상태, 업데이트가 필요한 상태 두 개로 구분됨

Related:
- https://github.com/next-step/spring-subway-service-kakao/pull/25#discussion_r565021613

Jjungs7 commented on src/main/java/subway/path/domain/SubwayGraph.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.function.Supplier;
+
+public class SubwayGraph {
+    private final Supplier<List<SectionEdge>> graphUpdater;
```
제가 supplier를 쓴 이유가 
```java
subwayGraph.update(sectionDao.getSectionEdges());
```
와 같이 할 경우 쿼리를 날려서 캐싱하지 않게 되고, supplier를 쓰지 않으려면 아래와 같이 state를 조회해서 어떤 상태인지 확인해야 되더라구요. sectionDao가 맡아야 할 일을 subwayGraph에 저장하는 것 같아서 supplier를 쓰지 말아야 할 것 같은데 더 좋은 방법이 있을지 궁금합니다!
```java
if (subwayGraph.getState() instanceof OutOfDate) {
    subwayGraph.update(sectionDao.getSectionEdges());
}
```
이 외에도 스테이트 패턴을 의도하신 것에 맞게 사용했을까요??

htrucci replied:
고민을 많이해서 잘 구현해주신 것 같네요 😄 
저도 supplier를 알지못해서 찾아보았습니다. ^^;;  
사실 복잡하게는 생각안하고 Section에 변화가 생겨 갱신이 필요한지 State를 별도의 인터페이스 및 상태로 분리했으면 좋겠다. 
그리고 SubwayGraph 내부에서든 관리를 해주는 외부서비스에서든 해당 상태에 따라 갱신해주면 되겠다 싶었습니다. 그렇게 되면 적어주신 것과 같은 분기문도 물론 포함되구요. 그 분기 조차 쓰지 않으려면,, 딱히 더 좋은 방법이 생각나지 않네요ㅠ

Jjungs7 commented:
앗 제가 `@JsonCreator`를 사용해서 실행이나 테스트할 때
`Settings -> Build, Execution, Deployment -> Compiler -> Java Compiler -> Additional command line parameters`
에다가 `-parameters` 추가해주셔야돼요!

Jjungs7 committed:
refactor(favorite): 유저가 직접 즐겨찾기에 등록한 노선만 수정할 수 있음

Jjungs7 committed:
refactor(loginmember): 게스트 저장 위치 변경

htrucci commented on src/main/java/subway/auth/application/AuthService.java:
```diff
@@ -35,7 +35,7 @@ private boolean userExists(String email, String password) {
             return member.getEmail().equals(email) && member.getPassword().equals(password);
         } catch (DataAccessException e) {
-            throw new AuthorizationException(e.getMessage());
+            throw new UnauthenticatedException("Incorrect email or password\n" + e.getMessage());
```
👍

htrucci commented on src/main/java/subway/auth/ui/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -31,15 +33,17 @@ public boolean supportsParameter(MethodParameter parameter) {
+        String token = AuthorizationExtractor.extract(request);
+        if (jwtTokenProvider.validateToken(token)) {
+            return memberService.findMemberByEmail(jwtTokenProvider.getPayload(token));
         }
```
잘 바꿔주셨네요 👍

htrucci commented on src/main/java/subway/member/ui/MemberController.java:
```diff
@@ -26,24 +26,6 @@ public MemberController(MemberService memberService) {
         return ResponseEntity.created(URI.create("/members/" + member.getId())).build();
     }
 
-    @GetMapping("/members/{id}")
-    public ResponseEntity<MemberResponse> findMember(@PathVariable Long id) {
-        MemberResponse member = memberService.findMember(id);
-        return ResponseEntity.ok().body(member);
-    }
-
-    @PutMapping("/members/{id}")
-    public ResponseEntity<MemberResponse> updateMember(@PathVariable Long id, @RequestBody MemberRequest param) {
-        memberService.updateMember(id, param);
-        return ResponseEntity.ok().build();
-    }
-
-    @DeleteMapping("/members/{id}")
-    public ResponseEntity<MemberResponse> deleteMember(@PathVariable Long id) {
-        memberService.deleteMember(id);
-        return ResponseEntity.noContent().build();
-    }
```
사용하지 않아 지워도 상관없곘으나 updateMember, deleteMember가 어드민 기능이 아니고 내정보 수정 및 탈퇴 기능이라면 
동일하게 @AuthenticationPrincipal 토큰처리를 통해 자신일때만 수행이 가능하게 바꿀 수 있곘죠? 😄

htrucci reviewed (approved):
안녕하세요 이안!
피드백 모두 잘 수정해주셨네요. 
감사드리고, 추가로 의견드릴 부분이 없어 머지하도록 하겠습니다.
질문해주신 부분에 대한 답이 잘 되었을지 모르겠는데 
저도 추가로 더 고민해보고 다시 말씀드리겠습니다.😭 
미션 진행하느라 고생하셨습니다 👏
